service/raft/raft_rpc.cc

```diff
@@ -30,135 +32,99 @@ raft_rpc::raft_rpc(raft_state_machine& sm, netw::messaging_service& ms,
-        return ser::raft_rpc_verbs::send_raft_append_entries_reply(&_messaging, netw::msg_addr(_address_map.get_inet_address(id)), timeout(), _group_id, _server_id, id, reply)
-            .handle_exception([id] (std::exception_ptr ex) {
+template <typename Verb, typename Msg> void
+raft_rpc::one_way_rpc(std::experimental::source_location loc, raft::server_id id,
+        Verb&& verb, Msg&& msg) {
+    (void)with_gate(_shutdown_gate, [this, loc = std::move(loc), id, &verb, &msg] {
+        auto ip_addr = _address_map.find(id);
+        if (!ip_addr) {
+            rlogger.debug("{}:{}: {} dropping outgoing message to {} - IP address not found",
+                loc.file_name(), loc.line(), loc.function_name(), id);
+            return make_ready_future<>();
+        }
+        return verb(&_messaging, netw::msg_addr(*ip_addr), timeout(), _group_id, _server_id, id, std::forward<Msg>(msg))
+            .handle_exception([loc = std::move(loc), id] (std::exception_ptr ex) {
                 try {
                     std::rethrow_exception(ex);
                 } catch (seastar::rpc::timeout_error&) {
                 } catch (seastar::rpc::closed_error&) {
                 } catch (...) {
-                    rlogger.error("Failed to send append reply to {}: {}", id, std::current_exception());
+                    rlogger.error("Failed to send {} to {}: {}", loc.function_name(), id, std::current_exception());
                 }
-            });
+        });
     });
 }
 
-void raft_rpc::send_vote_request(raft::server_id id, const raft::vote_request& vote_request) {
-    (void)with_gate(_shutdown_gate, [this, id, vote_request] {
-        return ser::raft_rpc_verbs::send_raft_vote_request(&_messaging, netw::msg_addr(_address_map.get_inet_address(id)), timeout(), _group_id, _server_id, id, vote_request)
-            .handle_exception([id] (std::exception_ptr ex) {
-                try {
-                    std::rethrow_exception(ex);
-                } catch (seastar::rpc::timeout_error&) {
-                } catch (seastar::rpc::closed_error&) {
-                } catch (...) {
-                    rlogger.error("Failed to send vote request {}: {}", id, ex);
-                }
-            });
+template <typename Verb, typename... Args>
+decltype(auto)
+raft_rpc::two_way_rpc(const std::experimental::source_location loc, raft::server_id id,
+        Verb&& verb, Args&&... args) {
+    auto ip_addr = _address_map.find(id);
+    using Fut = decltype(verb(&_messaging, netw::msg_addr(*ip_addr), db::no_timeout, _group_id, _server_id, id, std::forward<Args>(args)...));
+    using Ret = typename Fut::value_type;
+    if (!ip_addr) {
+        const auto msg = format("Failed to send {} {}: ip address not found", loc.function_name(), id);
+        return make_exception_future<Ret>(raft::transport_error(msg));
+    }
+    return verb(&_messaging, netw::msg_addr(*ip_addr), db::no_timeout, _group_id, _server_id, id, std::forward<Args>(args)...)
```

Comment:
I believe because you can't really abort a two-way RPC in our RPC impl today. I.e. you can pretend it's aborted, but there has to be a future somewhere which reads the result. So it doesn't make sense. @gleb-cloudius I remember we discussed it with you some time ago.